### ISSUE TITLE
Cache client-side static files using Service Workers

### ISSUE THREAD

**Issue by alshakero** [Polyjuice/SysUpgrade#111]
This is important for app performance and UX. And it's key for an app-like experience.

I'll do it in three steps:

- [x] Continue investigating around https://github.com/Starcounter/RebelsLounge/issues/273 to create the best way that lists all static files that works on Windows
- [x] Create a service worker that consumes the list generated by first point
- [x] Create a service worker that caches in runtime.
- [ ] Make sure it works on iOS and works as an app from Homescreen.

**Comment by warpech** [comment]
The proof-of-concept PR in https://github.com/Starcounter/UniformDocs/pull/357 is almost finished. We will soon make PRs in StarcounterClientFiles and Level1.

Before a docs page is created, this GitHub issue needs to host answers for the below questions:

Things that must be noted in the docs:

0. primer on what is a Service Worker, and where to learn more
   - what browsers are supported, desktop and mobile. Make sure we test in these browsers manually at least once after the alpha release.
   - what is the cache size
1. the caching strategy that we have implemented (runtime cache, not precache, single key, invalidated when any static file changes). Is this one of the strategies listed in https://serviceworke.rs/caching-strategies.html, or why not?
    - include conclusions from https://github.com/Starcounter/UniformDocs/pull/357#issuecomment-440363411. Explain what exactly happens to the tabs that are opened at the time the static resource is changed.
2. what kinds of resources are cached
    - why the URI is at the root directory (https://github.com/Starcounter/UniformDocs/pull/357#discussion_r215575992)
3. ways to reset the cache (change a file on disk, restart the database or an app, anything else? This comment has a nice summary https://github.com/Starcounter/UniformDocs/pull/357#issuecomment-484881432, also here: https://github.com/Starcounter/UniformDocs/pull/357#issuecomment-421926315)
4. how to opt out a resource from caching
5. what are diagnostic console messages and when to expect them
6. how to C# API for cache invalidation works
7. how to disable service worker and uninstall installed sw's (https://github.com/Starcounter/UniformDocs/pull/357#issuecomment-485549246)
8. how it gracefully fails in HTTP when the host is not `localhost` (https://github.com/Starcounter/UniformDocs/pull/357/files/8202265524e5003613f2316d0016e50b669fe088#diff-39a770e183ba0530a4afafd3c2f5d8d7)
9. information that "Disable cache" and flushing the cache in Chrome DevTools does not disable nor flush SW and its caches. Information how to disable SW.
10. how does SW behave in a multihost app served from the same Starcounter database. Is the cache shared between the hosts?

**Comment by alshakero** [comment]
Comment removed from app-shell and to be added to docs:

```
In production, we register a service worker to serve assets from local cache.
This lets the app load faster on subsequent visits in production, and gives
it offline capabilities. However, it also means that developers (and users)
will only see deployed updates on the "N+1" visit to a page, since previously
cached resources are updated in the background. 

It's worth noting this the N+1 visit happens automatically after visit N, because
the logic below reloads the page when it detects changes in service workers cache key

To learn more about the benefits of this model, read https://goo.gl/KwvDNy
This link also includes instructions on opting out of this behavior.
```

**Comment by warpech** [comment]
While we are waiting for fixes in https://github.com/Starcounter/level1/pull/4929, @alshakero could you draft a documentation page about SW using the information gathered above? Can be done as a comment in this issue.

**Comment by alshakero** [comment]
Draft is here https://github.com/Starcounter/StarcounterClientFiles/releases/tag/untagged-c2125c7cba128270e39f

**Comment by warpech** [comment]
Thanks, I copied to the actual release 3.26.0 which was already there and deleted the draft.

**Comment by warpech** [comment]
Reassigning to me, I will finish this issue asap.

**Comment by warpech** [comment]
Released in StarcounterClientFiles 3.26.0 and Level1 https://github.com/Starcounter/level1/commit/f7556758c8bdffc6a210a536de31fce54f0bebb4.

Level1 build: https://teamcity.starcounter.org/viewLog.html?buildId=96664&tab=buildResultsDiv&buildTypeId=ScMappingNightlyRel

Release notes: https://github.com/Starcounter/StarcounterClientFiles/releases/tag/3.26.0

Part of the release notes will be included in Starcounter docs.
